### DISCUSSION TITLE
wire:ignore possible issue when RTE is removed?

### DISCUSSION THREAD

**Discussion by Fredericld** [livewire/livewire#3094] [7. Bugs (archived)]
### Description
Hi

I am a bit at a loss here. This may be an issue, many be not. I can not quite understand what's going on.
I have built a form allowing me to add as many as Tinymce instances as I need.

The HTML generating the form is as follow
```
        <div id="questions" class="tab-pane">
            <div class="row">
                <div class="col-lg-6">

                    @foreach($relatedQuestions as $key => $relatedQuestion)

                        <div class="form-group" >
                            
                            <div wire:ignore>

                                {!! Form::textarea('relatedQuestions['.$key.'][title]', '', array( 'placeholder' => 'Question','class' => 'form-control tiny_question_title', 'wire:model.lazy' => 'relatedQuestions.'.$key.'.title', 'id' => 'relatedQuestions['.$key.'][title]' )) !!}
                                @error('relatedQuestions.'.$key.'.title')<span class="text-danger error">{{ $message }}</span>@enderror
                            </div>

                            <br/>
                            <input type="textarea" class="form-control tiny_question_answer" placeholder="Enter text"  name="relatedQuestions[{{$key}}]['text']" wire:model.lazy="relatedQuestions.{{$key}}.text">
                            @error('relatedQuestions.'.$key.'.text')<span class="text-danger error">{{ $message }}</span>@enderror
                        </div>

                        <button class="btn btn-danger btn-sm" wire:click.prevent="removeRelatedQuestion({{ $key }});">remove question {{ $key }}</button>

                    @endforeach

                    <button class="btn text-white btn-info btn-sm" wire:click.prevent="addRelatedQuestion();">Add a question</button>

                </div>
            </div>
        </div>
```

```
    window.addEventListener('componentUpdated', event => {
        addTinyMCE();
    });


    // Add TinyMCE
    function addTinyMCE(){
        tinymce.init({
            selector: '.tiny_question_title',
            plugins: [
                'advlist autolink link lists charmap print preview hr anchor pagebreak spellchecker',
                'searchreplace wordcount visualblocks visualchars code fullscreen insertdatetime media image nonbreaking',
                'save table directionality emoticons template paste'
            ],
            relative_urls: true,
            document_base_url: '{{ Config::get('app.url') }}',
            file_picker_callback (callback, value, meta) {
                let x = window.innerWidth || document.documentElement.clientWidth || document.getElementsByTagName('body')[0].clientWidth
                let y = window.innerHeight|| document.documentElement.clientHeight|| document.getElementsByTagName('body')[0].clientHeight

                tinymce.activeEditor.windowManager.openUrl({
                url : '/file-manager/tinymce5',
                title : 'Laravel File manager',
                width : x * 0.8,
                height : y * 0.8,
                onMessage: (api, message) => {
                    callback(message.content, { text: message.text })
                }
                })
            },
            setup: function(editor) {
                editor.on('init change', function () {
                    editor.save();
                });
                editor.on('blur', function(e) {
                    myStr = tinymce.activeEditor.id;
                    id = myStr.match(/\d+/);
                    @this.set('relatedQuestions.'+id+'.title', tinymce.get('relatedQuestions['+id+'][title]').getContent());
                });
            }

        });

        
    }

```

I can add as many TinyMCE to the screen as I want, I can save & load them to the DB.

The issue I have is I cannot remove a question. Every time, I use the "remove" button, it removes the correct element in the PHP but on screen, it always remove the last insert TinyMCE instance.

```
    /**
     * Add a question
     */
    public function addRelatedQuestion() 
    {
        $this->relatedQuestions[] = ['title' => ''];

        //converts the textarea to timymce
        $this->dispatchBrowserEvent('componentUpdated');

    }


/**
     * Remove a question
     */
    public function removeRelatedQuestion($id)
    {
        unset($this->relatedQuestions[$id]);
    }
```

The PHP seems correct but It feels the wire:ignore of livewire is not quite working and prevents the deletion of the tinyMCE editor that should be removed.

**Context**
- Livewire version: [e.g. 2.2.3]
- Laravel version: [e.g. 8.3.1]
- Browser: Chrome

**Comment by ju5t** [comment]
`wire:ignore` is a property that tells Livewire to ignore the element. The screencast on [wire:ignore](https://laravel-livewire.com/screencasts/s4-wire-ignore) has a good explanation. 

In short, everything inside an element that has `wire:ignore` will not be updated when Livewire 'refreshes' the page.

This is not a bug. This is as it is intended. Why are you using it?

**Comment by Fredericld** [comment]
I need to add a questions and answers to a page.
My only issue is when I remove a question/answer, the tinymce do not refresh because the wire:ignore. How can I force it to refresh if I do a delete?

**Comment by ju5t** [comment]
You can't. You have to remove `wire:ignore`.

**Comment by joshhanley** [comment]
Generally you need wire:ignore to get other libraries to run as they insert content into the DOM which Livewire deletes.

@Fredericld but you won't be able to use Livewire to update the text editor due to the wire:ignore. You will have to do that using javascript. If you have a look at your libraries docs you will probably find javascript methods or hooks that you can use to update the content. 

The easiest solution would be to use Alpine to watch for changes with Livewire and call any functions on the text editor to push the changes into the editor, and vice-versa, pick up any editor changes with Alpine and send through to Livewire.

Hope this helps!

**Comment by joshhanley** [comment]
@Fredericld sorry, I just read through the whole issue, and I was just replying to the comments before, not the actual problem.

I can see you already have the javascript setup.

The problem with deleting will be a morphdom issue. You will need to make sure you put a `wire:key` on any root level elements inside the loop, so Livewire knows to delete the correct one.

I would wrap everything that is inside `@foreach` and `@endforeach` in a `div` so there is one root element inside the loop. 

Then on that div put a key like 
```html
<div wire:key="related-question-{{$key}}">
````
 Or if the question has an ID that would be even better as it's unique, if that's the cause use the question id instead of the $key `wire:key="related-question-{{$relatedQuestion->id}}"`

**Comment by Fredericld** [comment]
@joshhanley 
Thanks Josh. The wire-key has definitely helped as the correct tinyMCE gets deleted. 
A new issue has now arised ...  Some of the tinyMCE in the page become unresponsive as I can not type anything in them. Their text panes become empty but I can see via the 'source code' button of the editors that they contain the correct text.
Some of the editors behave properly though...

Each editor has its own 32 characters unique ID

for example if I have 7 editors
1
2
3
4
5
6
7

delete the first one
I now get:
2 (becomes unresponsive)
3 (becomes unresponsive)
4
5
6
7

Any help or suggestions would be very appreciated....

**Comment by joshhanley** [comment]
@Fredericld hmm hard to say without being about to test it. I suspect that even with the correct keys that they might be getting reinitialised (or trying to) or something. Any chance you could put a test repo together that demonstrates the issue, so I can run it locally and have a look?

**Comment by ju5t** [comment]
@Fredericld in the [discussion](https://github.com/livewire/livewire/discussions/2273) you opened I suggested using blade components. This does not only make adding in a TinyMCE component easier, it also ensures each component has its own scope. 

I just spent some time trying to come up with a solution based on components; I'm getting close. I can now add a TinyMCE editor and link it to every model that you need. For example:

```blade
<x-input.tinymce wire:model="editor" placeholder="Type anything you want..." />
```

There still is one issue. When I run this on blur from within the editor:

```js
editor.on('blur', function(e) {
    value = editor.getContent();
})
```

The editor is completely removed. 

```js
value = tinymce.activeEditor.getContent();
```

Has the same error.

I have never worked with TinyMCE and it's getting late so I'm wrapping up. If you know why this is happning, I can finish it and send it here to you.

**Comment by ju5t** [comment]
Got it. Had `wire:ignore` on the wrong element. I will try to make things a little cleaner tomorrow and will post it here.

**Comment by ju5t** [comment]
This is the TinyMCE component: https://gist.github.com/ju5t/579a90b03f5cb80f3c6984d48a3385c4. There are some instructions in this gist too.

I can't promise it will solve your issues, but there might be a chance that it will as all logic is scoped. Every time you add the component with `x-input.tinymce` all of the JavaScript it holds is scoped and will not leave the component.

**Comment by ju5t** [comment]
@Fredericld out of curiosity, did my suggestion with a blade component work?

**Comment by Fredericld** [comment]
@ju5t 
I have not been able to try it out yet. I reviewed your code last week and cannot wait to give it a go. 
I have had to switch to another section in the project but I will get back to you when i go back to the tinymce section. Thanks for the help by the way!

**Comment by Fredericld** [comment]
Hi

I have been able to try out your code and I get the same result with your code and my code. Somehow the body of some of the tinymce are lost after a delete
I created a a sample of the code I use here: https://laravelplayground.com/#/snippets/25d9cdf3-c36f-4f12-92f1-f074ba6fc3ed
It does not fully works as their environment as tehey prevent the textarea from becoming RTEs.  Bbut if you load that code in your own system you should see the issue. The tinymces will become unresponsive and lose their body content after a delete.

Playground does not allow me to add folders so I was unable to add your code there.

**Comment by ju5t** [comment]
This is a whole lot trickier than I thought. It's not really related to Livewire. There is an issue on [Stack Overflow](https://stackoverflow.com/questions/2535569/tinymce-editor-dislikes-being-moved-around) that covers this particular issue. Livewire will move the editors when you remove them. At that point, TinyMCE loses it.

In the end I don't think there is a clean solution. The only thing I can come up with is to remove the editor and add it back after Livewire has rendered the component. This will flicker the editor (as in, you will most likely notice it) though and definitely isn't clean.

To point you in the right direction:

```javascript
document.addEventListener("DOMContentLoaded", () => {
    Livewire.hook('element.updating', (fromEl, toEl, component) => {
        for (instance in tinymce.editors) {
            tinymce.editors[instance].remove()
        }
    })
    Livewire.hook('element.updated', (el, component) => {
        // Enable here
    })
});
```

This should go on the page that includes the loop. If you put in in a `script`-tag you're nearly there. You do need to enable TinyMCE again though. Not sure if there is a way without removing TinyMCE, perhaps disabling and enabling is sufficient.

I'm sorry but this goes beyond what I can do. I don't know enough about TinyMCE and I'm not a fan it. We will never use it :-)
